src/watchdog.js

```diff
@@ -0,0 +1,414 @@
+		this._config = cloneDeepWith( config, function leaveDOMReferences( value ) {
+			return isElement( value ) ? value : undefined;
+		} );
+
+		return Promise.resolve()
+			.then( () => this._creator( elementOrData, this._config ) )
+			.then( editor => {
+				this._editor = editor;
+
+				window.addEventListener( 'error', this._boundErrorHandler );
+				this.listenTo( editor.model.document, 'change:data', this._throttledSave );
+
+				this._lastDocumentVersion = editor.model.document.version;
+				this._data = editor.data.get();
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Restarts the editor instance. This method is also called whenever an editor error occurs.
+	 * It fires the `restart` event.
+	 *
+	 * @fires restart
+	 * @returns {Promise.<module:core/editor/editor~Editor>}
+	 */
+	restart() {
+		this._throttledSave.flush();
+
+		return Promise.resolve()
+			.then( () => this.destroy() )
+			.catch( err => console.error( 'An error happened during the editor destructing.', err ) )
+			.then( () => {
+				if ( typeof this._elementOrData === 'string' ) {
+					return this.create( this._data, this._config );
+				}
+
+				const updatedConfig = Object.assign( {}, this._config, {
+					initialData: this._data
+				} );
+
+				return this.create( this._elementOrData, updatedConfig );
+			} )
+			.then( () => {
+				this.fire( 'restart' );
+			} );
+	}
+
+	/**
+	 * Destroys the current editor using the destructor passed to {@link #setDestructor} method.
+	 *
+	 * @returns {Promise.<module:watchdog/watchdog~Watchdog>}
+	 */
+	destroy() {
+		window.removeEventListener( 'error', this._boundErrorHandler );
+		this.stopListening( this._editor.model.document, 'change:data', this._throttledSave );
+
+		return Promise.resolve()
+			.then( () => this._destructor( this._editor ) )
+			.then( () => {
+				this._editor = null;
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Saves the editor data, so it can be restored after the crash even if the data cannot be fetched at
+	 * the moment of a crash.
+	 *
+	 * @private
+	 */
+	_save() {
+		const version = this._editor.model.document.version;
+
+		// Change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( version < this._lastDocumentVersion ) {
+			this._throttledSave.cancel();
+
+			return;
+		}
+
+		try {
+			this._data = this._editor.data.get();
+			this._lastDocumentVersion = version;
+		} catch ( err ) {
+			console.error(
+				err,
+				'An error happened during restoring editor data. ' +
+				'Editor will be restored from the previously saved data.'
+			);
+		}
+	}
+
+	/**
+	 * Checks if the event error comes from the editor that is handled by the watchdog (by checking the error context) and
+	 * restarts the editor. It handles {@link module:utils/ckeditorerror~CKEditorError CKEditorError errors} only.
+	 *
+	 * @private
+	 * @fires error
+	 * @param {Event} event Error event.
+	 */
+	_handleGlobalErrorEvent( event ) {
```

Comment:
We usually use `evt`, not `event`.

Reply:
We use `evt` for our custom event. Here this is a "real" event

Reply:
That's true, but I already changed it. Hope it's fine.

Reply:
I'm fine with the other name too 😄